storage/jellyfish_merkle/src/restore/mod.rs

```diff
@@ -0,0 +1,411 @@
+                        let mut internal_info = InternalInfo::new_uninitialized(new_node_key);
+
+                        // Next we put the existing leaf as a child of this internal node.
+                        let existing_child_index =
+                            existing_leaf.account_key().get_nibble(common_prefix_len);
+                        internal_info.set_child(
+                            existing_child_index as usize,
+                            ChildInfo::Leaf {
+                                node: existing_leaf,
+                            },
+                        );
+
+                        // Do not set the new child for now. Do it later after freezing the
+                        // previous leaf.
+                        self.partial_nodes.push(internal_info);
+                        self.freeze(self.partial_nodes.len());
+
+                        // Now we set the new child.
+                        let new_child_index = new_key.get_nibble(common_prefix_len);
+                        self.partial_nodes
+                            .last_mut()
+                            .expect("This node must exist.")
+                            .set_child(
+                                new_child_index as usize,
+                                ChildInfo::Leaf {
+                                    node: LeafNode::new(new_key, new_value),
+                                },
+                            );
+
+                        break;
+                    }
+                }
+                None => {
+                    // This means that we are going to put a leaf in this position. For all the
+                    // descendants on the left, they are now frozen.
+                    self.freeze(i + 1);
+
+                    // Mark this position as a leaf child.
+                    self.partial_nodes[i].set_child(
+                        child_index,
+                        ChildInfo::Leaf {
+                            node: LeafNode::new(new_key, new_value),
+                        },
+                    );
+
+                    // We do not add this leaf node to self.frozen_nodes because we don't know its
+                    // node key yet. We will know its node key when the next account comes.
+                    break;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Puts the nodes that will not be changed in `self.frozen_nodes`.
+    fn freeze(&mut self, target_len: usize) {
+        // Freeze the rightmost leaf node on the lowest level. This node was inserted in the
+        // previous `restore_one` call.
+        let last_node = self
+            .partial_nodes
+            .last()
+            .expect("Must have at least one partial node.");
+        for i in (0..16).rev() {
+            if let Some(ref child_info) = last_node.children[i] {
+                if let ChildInfo::Leaf { node } = child_info {
```

Comment:
Maybe match on ChildInfo instead so you can handle the ChildInfo::Internal case with either assert! or unreachable!